services/wallet/currency/currency.go

```diff
@@ -0,0 +1,150 @@
+package currency
+
+import (
+	"errors"
+	"math"
+	"strings"
+
+	iso4217 "github.com/ladydascalie/currency"
+
+	"github.com/status-im/status-go/services/wallet/price"
+	"github.com/status-im/status-go/services/wallet/token"
+)
+
+const decimalsCalculationCurrency = "USD"
+
+type Format struct {
+	Symbol              string `json:"symbol"`
+	DisplayDecimals     uint   `json:"displayDecimals"`
+	StripTrailingZeroes bool   `json:"stripTrailingZeroes"`
+}
+
+type FormatPerSymbol = map[string]Format
+
+type Currency struct {
+	priceManager *price.Manager
+}
+
+func NewCurrency(priceManager *price.Manager) *Currency {
+	return &Currency{
+		priceManager: priceManager,
+	}
+}
+
+func IsCurrencyFiat(symbol string) bool {
+	return iso4217.Valid(strings.ToUpper(symbol))
+}
+
+func GetAllFiatCurrencySymbols() []string {
+	return iso4217.ValidCodes
+}
+
+func calculateFiatDisplayDecimals(symbol string) (uint, error) {
+	currency, err := iso4217.Get(strings.ToUpper(symbol))
+
+	if err != nil {
+		return 0, err
+	}
+
+	return uint(currency.MinorUnits()), nil
+}
+
+func calculateFiatCurrencyFormat(symbol string) (*Format, error) {
+	displayDecimals, err := calculateFiatDisplayDecimals(symbol)
+
+	if err != nil {
+		return nil, err
+	}
+
+	format := &Format{
+		Symbol:              symbol,
+		DisplayDecimals:     displayDecimals,
+		StripTrailingZeroes: false,
+	}
+
+	return format, nil
+}
+
+func calculateTokenDisplayDecimals(price float64) uint {
+	var displayDecimals float64 = 0.0
+
+	if price > 0 {
+		const lowerCurrencyResolution = 0.1
+		const higherCurrencyResolution = 0.01
+		lowerDecimalsBound := math.Max(0.0, math.Log10(price)-math.Log10(lowerCurrencyResolution))
+		upperDecimalsBound := math.Max(0.0, math.Log10(price)-math.Log10(higherCurrencyResolution))
```

Comment:
> I'm not sure I understand the logic behind this, can you please clarify?
This is supposed to reflect the specification (https://github.com/status-im/status-desktop/issues/8640).
You want as many decimal places as needed so that the last decimal place is worth between 1 and 10 USD cents.

If you've got N decimals available, the value of your last decimal place (your resolution) is `10ˆ(-N)`
Example: 2 decimals gives you a resolution of 0.01, 3 decimals gives you 0.001

That's in token units, so how much is your last decimal place worth in USD? If `P` is your price in USD, your resolution is worth `P * 10ˆ(-N)`

You want your resolution to be worth `Z` USD, then:
`Z = P * 10ˆ(-N)`
`log10(Z) = log10(P * 10ˆ(-N))`
`log10(Z) = log10(P) + log10(10ˆ(-N))`
`log10(Z) = log10(P) - N`
`N = log10(P) - log10(Z)`

> Also why not use +1 and +2 instead of -math.Log10(...)?
0.1 and 0.01 (amount in USD) are the limits set by the specification, which I guess could change if John changes his mind in the future. That's why I decided to leave the calculation in place instead of using the pre-calculated results of those `Log` of those values. I'm gonna assume the compile is smart enough to do that optimization for me.

Reply:
Thanks, @dlipicar, for the great clarifications.

For configuration values, like `lowerCurrencyResolution` and `higherCurrencyResolution`, I, personally, like having them at the top of the code with comments for easy finding not deep down in the code, not sure if this is the case.

Reply:
You're right... Changed the location and gave it a clearer name